### ISSUE TITLE
No screencasting with audio on Linux

### ISSUE THREAD

**Issue by damnkrat** [telegramdesktop/tdesktop#26642]
### Steps to reproduce

1. Try sharing video
2. There is no share audio button at all

### Expected behaviour

A button to share audio

### Actual behaviour

No ability to share audio at all. No matter X11 or Wayland

### Operating system

Linux

### Version of Telegram Desktop

4.8.7

### Installation source

Other (unofficial) source -> From Manjaro repo

**Comment by damnkrat** [comment]
@Aokromes  Can you provide a link to the duplicate? I couldn't related issue before creating new one

**Comment by ilya-fedin** [comment]
I don't think this was reported before. Although there's no contributor who can implement that so it's unlikely to appear...

**Comment by Aokromes** [comment]
it was, i remember it reported, but hard to find xd

**Comment by ilya-fedin** [comment]
I see there were questions about why this is not implemented (#16730, #17163) but not bug report about that. Although I'd say this should be a feature request.

**Comment by damnkrat** [comment]
Also both are marked as closed, we need to open one of the questions, or without that, do this feature

**Comment by ilya-fedin** [comment]
I don't see a reason to open those questions, they're answered

**Comment by ilya-fedin** [comment]
I'm highly doubt the feature will be implemented, with an issue or without. But you can try to open a new feature request, hope @Aokromes won't close anymore.

**Comment by ReillyBrogan** [comment]
I assume Telegram is using the XDG portal for screencasting? If so then sending audio with application/system capture is currently unsupported (which is why it doesn't work in Discord or OBS either). There has been some movement on this recently, I'd recommend keeping watch on [this](https://github.com/flatpak/xdg-desktop-portal/issues/957) issue (and for the Telegram devs working on Linux it might be a good idea to contribute their own two-cents there).

**Comment by ilya-fedin** [comment]
> I assume Telegram is using the XDG portal for screencasting?

WebRTC.

> There has been some movement on this recently, I'd recommend keeping watch on [this](https://github.com/flatpak/xdg-desktop-portal/issues/957) issue (and for the Telegram devs working on Linux it might be a good idea to contribute their own two-cents there).

I'm not sure that would help. WebRTC has no "capture system audio" API and if this would become a part of the screencasting API then it would become unimplementable for tdesktop as currently the portal session is managed by WebRTC and tdesktop doesn't have access to its handles so it would be unable to speak to the portal directly and WebRTC doesn't have an API for this.

Right now it's implementable as application has full access to PipeWire/PulseAudio/ALSA. The only problem is application uses OpenAL for audio and OpenAL doesn't have API for this feature either so this feature has to be implemented with native code while tdesktop project has no one with PipeWire/PulseAudio/ALSA experience to implement any backend (not even talking about every) for Linux. I guess OBS and Discord having the same problem: they don't have experience with Linux audio APIs and they always have much more prioritized tasks to spend time on rather than learning the zoo of Linux audio APIs.

I know there's https://github.com/maltejur/discord-screenaudio/blob/master/src/virtmic.cpp which could be used as a base but still someone has to spend time to port the implementation to tdesktop...

**Comment by gmankab** [comment]
some additional info about screen sharing with audio on wayland + pipewire

- works by default in [obs](https://github.com/obsproject/obs-studio)
- works by default in [discord-screenaudio](https://github.com/maltejur/discord-screenaudio)
- here is a [guide](https://github.com/IceDBorn/pipewire-screenaudio) how to make it work in firefox
- here is a [guide](https://github.com/edisionnano/Screenshare-with-audio-on-Discord-with-Linux) how to make it work in chromium

hope we get this feature in telegram soon

**Comment by ilya-fedin** [comment]
@gmankab someone interested should implement it and make a PR

**Comment by lucianosrp** [comment]
This feature is still not present as of v4.16.6 , would be nice to have it implemented somehow

**Comment by ilya-fedin** [comment]
No one seem to be really interested doing such work :(

**Comment by damnkrat** [comment]
For those who want some workaround for a time, check out helvum. Just connect audio sources manually in it

![Снимок экрана_20240418_154244](https://github.com/telegramdesktop/tdesktop/assets/32395839/0fc6a53c-6dfc-4314-bc9a-4b14a81759e1)
